### ISSUE TITLE
Compile error:  ‘__T0’ was not declared in this scope

### ISSUE THREAD

**Issue by stsievert** [pytorch/pytorch#3243]
When I build PyTorch from source following the [README directions][1], I get an error:

``` shell
[ 64%] Building CXX object gloo/CMakeFiles/gloo.dir/transport/tcp/device.cc.o
[ 68%] Building CXX object gloo/CMakeFiles/gloo.dir/transport/tcp/pair.cc.o
[ 72%] Linking CXX static library libgloo.a
[ 72%] Built target gloo
[ 76%] Building NVCC (Device) object gloo/CMakeFiles/gloo_cuda.dir/gloo_cuda_generated_cuda_private.cu.o
[ 80%] Building NVCC (Device) object gloo/CMakeFiles/gloo_cuda.dir/gloo_cuda_generated_cuda.cu.o
/home/ec2-user/pytorch/torch/lib/gloo/gloo/common/linux.h:26:26: error: ‘__T0’ was not declared in this scope
 const auto kPCIClass3D = PCIClassMatch{0x030200, 0xffff00};
                          ^
/home/ec2-user/pytorch/torch/lib/gloo/gloo/common/linux.h:29:31: error: ‘__T0’ was not declared in this scope
 const auto kPCIClassNetwork = PCIClassMatch{0x020000, 0xff0000};
                               ^
CMake Error at gloo_cuda_generated_cuda_private.cu.o.cmake:263 (message):
  Error generating file
  /home/ec2-user/pytorch/torch/lib/build/gloo/gloo/CMakeFiles/gloo_cuda.dir//./gloo_cuda_generated_cuda_private.cu.o


make[2]: *** [gloo/CMakeFiles/gloo_cuda.dir/gloo_cuda_generated_cuda_private.cu.o] Error 1
make[2]: *** Waiting for unfinished jobs....
/home/ec2-user/pytorch/torch/lib/gloo/gloo/common/linux.h:26:26: error: ‘__T0’ was not declared in this scope
 const auto kPCIClass3D = PCIClassMatch{0x030200, 0xffff00};
                          ^
/home/ec2-user/pytorch/torch/lib/gloo/gloo/common/linux.h:29:31: error: ‘__T0’ was not declared in this scope
 const auto kPCIClassNetwork = PCIClassMatch{0x020000, 0xff0000};
                               ^
CMake Error at gloo_cuda_generated_cuda.cu.o.cmake:263 (message):
  Error generating file
  /home/ec2-user/pytorch/torch/lib/build/gloo/gloo/CMakeFiles/gloo_cuda.dir//./gloo_cuda_generated_cuda.cu.o


make[2]: *** [gloo/CMakeFiles/gloo_cuda.dir/gloo_cuda_generated_cuda.cu.o] Error 1
make[1]: *** [gloo/CMakeFiles/gloo_cuda.dir/all] Error 2
make: *** [all] Error 2  
```

Machine details:

* nvcc release 7.5, V7.5.17
* gcc version 4.8.3 20140911 (Red Hat 4.8.3-9) (GCC)
* gcc version 4.8.3 20140911 (Red Hat 4.8.3-9) (GCC)
* Python 3.6.1 with Anaconda
* OS: Red Hat

I'm not sure if I have cuDNN. I am using Amazon's EC2 Deep Learning AMI, and they say they have cuDNN drivers: https://aws.amazon.com/amazon-ai/amis/

[1]:https://github.com/pytorch/pytorch#from-source

**Comment by iffsid** [comment]
The issue appears to be a mismatch between `gcc` and `cuda-sdk` versions.
See #3105  for a discussion of the same issue.

**Comment by soumith** [comment]
this is a NVCC 7.5 bug. I tried various versions of gcc with nvcc in a CENTOS6 env, and it wasn't fixed. The only workaround is to fix gloo or update nvcc to 8.0 or greater.
I fixed gloo in the current subtree, the error should not happen right now.
Fixed via: https://github.com/facebookincubator/gloo/pull/107
